### PR TITLE
Ports TGMC timer cooldowns (now supporting signals)

### PR DESCRIPTION
**Pull request by Rohesie**
Added it to a random timer as proof of concept.

This is less performant than storing `world.time` in a variable, but it's a simple, quick, easy and dirty way to add a cooldown to something. Just create an ID and you need do no more besides using the macros.

Helps with clarity and adding cooldowns on the fly.

### CODE REVIEW

File: code/datums/datum.dm [hunk 1]
```diff
@@ -49,6 +49,9 @@
 	/// A weak reference to another datum
 	var/datum/weakref/weak_reference
 
+	///Lazy associative list of currently active cooldowns.
```
Comment (Qustinnus) [inline code/datums/datum.dm:52]: Can u include the assoc key and value that'll be in this list?

File: code/__DEFINES/cooldowns.dm [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+
+//INDEXES
+#define COOLDOWN_BOX_DEPLOYMENT	"box deployment"
+
+//MACROS
+#define COOLDOWN_START(cd_source, cd_index, cd_time) LAZYSET(cd_source.cooldowns, cd_index, addtimer(CALLBACK(GLOBAL_PROC, /proc/end_cooldown, cd_source, cd_index), cd_time))
+
+#define COOLDOWN_CHECK(cd_source, cd_index) LAZYACCESS(cd_source.cooldowns, cd_index)
+
+#define COOLDOWN_END(cd_source, cd_index) LAZYREMOVE(cd_source.cooldowns, cd_index)
```
Comment (Qustinnus) [inline code/__DEFINES/cooldowns.dm:16]: Cooldown reset sounds better but that's preference I guess

**Review by ghost — APPROVED** [review]
Looks good, is the plan to enforce this to any cooldown (unless the performance cost is too impactful)?

**Comment by Rohesie** [comment]
I don't think there's need to enforce it. It's just a simple and easy way to add cooldowns, but it doesn't offer many advantages beyond that. If people want to use a different method that's fine, though on several cases I'd suggest this one.

File: code/datums/datum.dm (hunk 1 above)
Comment (Rohesie) [reply code/datums/datum.dm:52]: Something like cooldown index -> return of add_timer (usually the truthy value of -1 unless someone makes a stoppable timer version)?

Comment (Qustinnus) [reply code/datums/datum.dm:52]: i ment in the comment

Comment (Rohesie) [reply code/datums/datum.dm:52]: I know, but I don't understand what you are asking.
Maybe give an example?

Comment (Qustinnus) [reply code/datums/datum.dm:52]: add the structure of the list in the code comment so people looking at it know what the list contains without having to go hunting for it.

Comment (Rohesie) [reply code/datums/datum.dm]: Like this?

File: code/__DEFINES/cooldowns.dm (hunk 2 above)
Comment (Rohesie) [reply code/__DEFINES/cooldowns.dm:16]: A reset would be a different variable for a stoppable timer.

File: code/datums/datum.dm (hunk 1 above)
Comment (Qustinnus) [reply code/datums/datum.dm]: yes perfect

**Comment by Rohesie** [comment]
@willox , take a peek at the changes and see if they match what you suggested.
I see benefits in both systems, really.

**Comment by Rohesie** [comment]
Added a bit of functionality to them. Timer cooldowns will send signals, reporting the events of finishing in time or early (and if early, they'll report how much time they had left).

**Comment by MrStonedOne** [comment]
One thing to look at is /vg/'s cooldown system, its basically a list of cooldown_id => world.time and some macros iirc. so basically your system but without going through sstimer and incurring like 4 proc overheads, the only downside is it doesn't auto-cleanup, you could how ever manage this using a subsystem.

**Comment by Rohesie** [comment]
> 
> 
> One thing to look at is /vg/'s cooldown system, its basically a list of cooldown_id => world.time and some macros iirc. so basically your system but without going through sstimer and incurring like 4 proc overheads, the only downside is it doesn't auto-cleanup, you could how ever manage this using a subsystem.

Where is this list being stored, @MrStonedOne ? I'm not finding it in their code. Could you get me the variable name perhaps?

There's support for `world.time` cooldowns here, it's the last system on the file. Timers cooldowns are mostly useful for if you want to react to the end of it, that's why they send a signal. Wouldn't make a processing subsystem to replicate that behavior kind of making a second timer SS?

**Comment by Rohesie** [comment]
Waiting on review, not really stale.

**Pull request merged**